R/infection_processes.R

```diff
@@ -21,161 +21,282 @@ create_infection_process <- function(
+  prophylaxis <- rep(0, length(source_humans))
+  drug <- api$get_variable(human, variables$drug, source_humans)
+  medicated <- (drug > 0)
+  if (any(medicated)) {
+    drug <- drug[medicated]
+    drug_time <- api$get_variable(
+      human,
+      variables$drug_time,
+      source_humans[medicated]
     )
+    prophylaxis[medicated] <- dweibull(
+      api$get_timestep() - drug_time,
+      parameters$drug_prophylaxis_shape[drug],
+      parameters$drug_prophylaxis_scale[drug]
+    )
+  }
 
-    b <- blood_immunity(ib[source_humans], parameters)
+  source_humans[bernoulli(length(source_humans), b * (1 - prophylaxis))]
+}
 
-    infected_humans <- source_humans[bernoulli(length(source_humans), b)]
+calculate_clinical_infections <- function(api, human, variables, infections) {
+  ica <- api$get_variable(human, variables$ica, infections)
+  icm <- api$get_variable(human, variables$icm, infections)
+  parameters <- api$get_parameters()
 
-    ica <- api$get_variable(
+  if (length(infections) > 0) {
+    timestep <- api$get_timestep()
+    boost_immunity(
+      api,
       human,
       variables$ica,
-      infected_humans
+      infections,
+      ica,
+      variables$last_boosted_ica,
+      timestep,
+      parameters$uc
     )
+    boost_immunity(
+      api,
+      human,
+      variables$id,
+      infections,
+      api$get_variable(human, variables$id, infections),
+      variables$last_boosted_id,
+      timestep,
+      parameters$ud
+    )
+  }
 
-    iva <- api$get_variable(
+  phi <- clinical_immunity(ica, icm, parameters)
+  infections[bernoulli(length(infections), phi)]
+}
+
+update_severe_disease <- function(
+  api,
+  clinical_infections,
+  infection_age,
+  human,
+  variables,
+  infections
+  ) {
+  if (length(clinical_infections) > 0) {
+    parameters <- api$get_parameters()
+    iva <- api$get_variable(human, variables$iva, infections)
+    theta <- severe_immunity(
+      infection_age,
+      iva,
+      api$get_variable(human, variables$ivm, clinical_infections),
+      parameters
+    )
+    develop_severe <- bernoulli(length(clinical_infections), theta)
+    api$queue_variable_update(
+      human,
+      variables$is_severe,
+      develop_severe,
+      clinical_infections
+    )
+    boost_immunity(
+      api,
       human,
       variables$iva,
-      infected_humans
+      infections,
+      iva,
+      variables$last_boosted_iva,
+      api$get_timestep(),
+      parameters$uv
     )
+  }
+}
 
-    icm <- api$get_variable(human, variables$icm, infected_humans)
-    phi <- clinical_immunity(ica, icm, parameters)
-
-    develop_clinical <- bernoulli(length(infected_humans), phi)
+calculate_treated <- function(
+  api,
+  human,
+  states,
+  variables,
+  clinical_infections
+  ) {
+  parameters <- api$get_parameters()
+  seek_treatment <- bernoulli(length(clinical_infections), parameters$ft)
+  n_treat <- sum(seek_treatment)
+
+  drugs <- sample(
+    parameters$clinical_treatment_drugs,
```

Comment:
What type does `parameters$clinical_treatment_drugs` have? If numeric, you need to be careful of a terrible design decision that `sample` has...

Reply:
`clinical_treatment_drugs` is numeric.

Ahhh If it's of length 1 then it will sample between `1:clinical_treatment_drugs`. Oooook, I can test and deal with that. Thanks!

Reply:
the other way is to do

```
x[sample.int(x, ...)]
```

(see ?sample) which won't have that surprise in it, and is often a bit clearer than an `if/else`